Fix custom styles for emails in preview and in sent messages

merkushin opened the pull request:
Resolves #7459 

Some styling that comes by default with WordPress couldn't be applied to emails. 

The main restriction comes from Gmail that removes unsupported styling from emails in their web-app. For this reason, we disable some settings in this PR.

We address the issue with misstyling of elements due to lack or conflict of styles.

## Proposed Changes
* Load global styles for inline CSS processor.
* Disable default color palettes.
* Disable alignment for buttons and images.
* Remove some default from editor settings that affect the appearance of the email in the editor only.

## Testing Instructions

For testing real messages, use Gmail webapp as it has more restrictive rules for CSS comparing to desktop applications.

1. Go to Sensei LMS > Settings > Emails.
2. Enable the Course Completed email and then edit it.
3. Change colors for buttons, paragraphs.
4. Change text alignment.
5. Add an image.
6. Make sure you don't have controls to change alignment of the image or buttons.
7. Save and preview. Make sure you see your changes in the preview.
8. Take a look at the button and make sure it has the same padding as in the editor.
9. Complete the course as a student.
10. Check your mailbox and make sure that the Course Completed email reflects your changes.

## Screenshots

### Real email 
(You can see that Sensei logo and the image in the body are not displayed — that's because they refer to my local website.)

#### Gmail web-app
![CleanShot 2024-06-02 at 15 29 49@2x](https://github.com/Automattic/sensei/assets/329356/b4efe93b-43ff-4d86-b89a-e61b24c44c9c)

#### Mail App
![CleanShot 2024-06-02 at 15 30 24@2x](https://github.com/Automattic/sensei/assets/329356/534603b1-47a5-4f37-a77d-364a839493ad)

### Editor
![CleanShot 2024-06-02 at 15 27 31@2x](https://github.com/Automattic/sensei/assets/329356/8a1eb008-95ad-49c7-8e76-ffb871d0ec20)

### Preview
![CleanShot 2024-06-02 at 15 27 44@2x](https://github.com/Automattic/sensei/assets/329356/e4b20f53-3313-4495-895a-fc619ea5eb23)

## Pre-Merge Checklist
<!-- Complete applicable items on this checklist **before** merging. Items that are not applicable can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed. -->
- [x] PR title and description contain sufficient detail and accurately describe the changes
- [ ] Acceptance criteria is met
- [x] Decisions are publicly documented
- [x] Adheres to coding standards ([PHP](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/), [JavaScript](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/), [CSS](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/), [HTML](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/))
- [ ] All strings are translatable (without concatenation, handles plurals)
- [x] Follows our naming conventions (P6rkRX-4oA-p2)
- [ ] Hooks (p6rkRX-1uS-p2) and functions are documented
- [ ] New UIs are responsive and use a [mobile-first approach](https://zellwk.com/blog/how-to-write-mobile-first-css/)
- [ ] New UIs match the designs
- [ ] Different user privileges (admin, teacher, subscriber) are tested as appropriate
- [ ] Legacy courses (course without blocks) are tested
- [ ] Code is tested on the minimum supported PHP and WordPress versions
- [ ] User interface changes have been tested on the latest versions of Chrome, Firefox and Safari
- [ ] "Needs Documentation" label is added if this change requires updates to documentation
- [ ] Known issues are created as new GitHub issues

merkushin committed:
Fix custom styles for emails in preview and in sent messages

merkushin commented on includes/internal/emails/class-email-sender.php:
```diff
@@ -229,17 +229,17 @@ public function get_email_body( WP_Post $email_post, array $placeholders = [] ):
 	 * @return string
 	 */
-	private function replace_placeholders( string $string, array $placeholders ): string {
+	private function replace_placeholders( string $content, array $placeholders ): string {
```
PHPCS was complaining for the variable name.

Imran92 committed:
Merge branch 'trunk' into fix/email-styles

Imran92 reviewed (commented):
For some reason it's not working for me -

Editor:

<img width="811" alt="Screenshot 2024-04-23 at 4 30 02 PM" src="https://github.com/Automattic/sensei/assets/6820724/63cc11ed-13bc-4441-9ed0-0c8f39932cab">

Gmail:

<img width="798" alt="Screenshot 2024-04-23 at 4 31 42 PM" src="https://github.com/Automattic/sensei/assets/6820724/cbaa2239-343f-4a3b-ae80-5c6ec86d67c5">

Temp mail client:

<img width="688" alt="Screenshot 2024-04-23 at 4 32 20 PM" src="https://github.com/Automattic/sensei/assets/6820724/1e00659f-9049-46cf-a91b-bd361d3c2231">

merkushin commented:
@Imran92 Thanks for testing!

Could you tell me what are the versions of WordPress and Gutenberg (if installed), and what theme do you use on the test website?

m1r0 commented:
I've tested it and the colors worked, but the alignment didn't. The result was the same with and without the Gutenberg plugin.

Imran92 commented:
> Could you tell me what are the versions of WordPress and Gutenberg (if installed), and what theme do you use on the test website?

Hi @merkushin 👋 I've shared the creds of a JN site where I reproduced the issue in DM. Hope it'll help! Thanks! Locally I didn't have GB installed and WP version was 6.5.2

merkushin commented:
Thanks Imran!

Ah, yes, the same email in Gmail and in the Mail app look so differently.

![CleanShot 2024-04-23 at 23 04 06@2x](https://github.com/Automattic/sensei/assets/329356/fda1dd7e-92b2-4701-a96b-4a13adce7c06)

![CleanShot 2024-04-23 at 23 03 52@2x](https://github.com/Automattic/sensei/assets/329356/c16cef13-5da0-4ff1-83da-79dc61e30e54)

merkushin committed:
Disable default colors in the editor for emails

merkushin committed:
Add styles for text and buttons alignment

merkushin commented:
As Imran suggested, I checked that when we don't use default colors and the picker instead we get colors as the HEX value, so I disabled default colors in the editor settings.
https://github.com/Automattic/sensei/pull/7595/commits/2a2ec4e10a8faa34e5b5eaf253acb287cb3eae7a

With alignment is a bit harder. I added classes needed we missed in CSS, now it works in Preview and in Mail app. But Gmail webapp "doesn't work". I found they just [don't support some CSS properties](https://developers.google.com/gmail/design/css#supported_css) (for example, `justify-content` and `align-items`—both are used to align buttons).
https://github.com/Automattic/sensei/pull/7595/commits/b8a6aabb23cf44688ddf9d78ddf06ecb55d9d5cb
And it is impossible to disable in the editor settings.

I also added styles to make text alignment work. 

I still see other issues with styles in emails:
- The size of the button is different in Editor vs Preview/Real Message.
- Image alignment doesn't work. I tried to fix it quickly by bringing needed classes, but it still misses the width property (not sure if it is the only cause), so it doesn't work.

And I suppose we can find even more tiny issues like those ones. I think we need to add a warning to our documentation that not all styles are supported in emails. And maybe list the ones we know already.

merkushin committed:
Merge branch 'trunk' into fix/email-styles

donnapep commented:
@merkushin @Imran92 What's the status of this one?  Pending another review or something else?

Imran92 commented:
> @merkushin @Imran92 What's the status of this one? Pending another review or something else?

I'll take a look at the last changes Dmitry made today!

Imran92 commented:
> As Imran suggested, I checked that when we don't use default colors and the picker instead we get colors as the HEX value, so I disabled default colors in the editor settings.

Glad it helped!

Can you please update the "Proposed Changes" and the "Testing Instruction" of the PR body to point out what changes are made, what to test and what are expected not to work yet?

Thanks!

Imran92 commented:
> And it is impossible to disable in the editor settings.

I wrote the following code to remove alignment settings from a block in the editor -

```
wp.hooks.addFilter(
    'blocks.registerBlockType',
    'imran/imranFilterName',
    removeJustificationAndAlignmentsFromButtons
);

function removeJustificationAndAlignmentsFromButtons(settings, name) {
    if (name !== 'core/buttons') {
        return settings;
    }

    return lodash.assign({}, settings, {
        supports: lodash.assign({}, settings.supports, {
            layout: false,
        }),
    });
}
```

Can you try and see if it works for you?

Without the script-

<img width="863" alt="image" src="https://github.com/Automattic/sensei/assets/6820724/8bfcd323-a25b-47fc-991b-9909a36c67c5">

With the script-
<img width="863" alt="image" src="https://github.com/Automattic/sensei/assets/6820724/179cb830-0279-4a72-bb7b-a8aede7dcdf3">

This is just an example, it should be technically possible to remove all types of _support_ this way. But there can be other issues which I haven't faced here but you did. Also, this code is just a quick one, proper syntax should be used for production.

Overall, most settings are working now. I've seen some issues as well.

- Button has _padding_ setting but it doesn't work
- Some color changes do not work

https://github.com/Automattic/sensei/assets/6820724/85b5e8c8-d623-4e11-9e66-249cb6c2ee0a

merkushin committed:
Disable alignment for buttons and images, fix padding for buttons and text color for paragraphs

merkushin commented:
Thanks @Imran92! Applied your suggestion to remove alignment for buttons and images.

I updated the PR description. The PR is ready for review.

Here are the details of the most recent changes:

With padding, there are the following styles that affect it:
- we have block-library/styles.css with `calc` for padding
- we have editor styles that use CSS variables for padding

Neither of them are loaded in the preview or for the actual email, and, basically, loading them doesn't make much sense if our goal is to look good in the Gmail web-app.

We also have our custom padding styles. By making the rule there `!important` we can make the appearance consistent. However, it makes impossible to apply custom padding in the editor.

While inspecting the editor settings, I found there is an array of styles. Setting it the an empty array removes a lot of unexpected (for us) styles and makes it possible to use padding.

It also makes alignment not working even in the editor. Which is good for us, I think, because there is a consistency between the editor and the preview. As I said in the beginning of the comment, the alignment feature was disabled anyway.

Also, I found the rule that defined the text color (that didn't work in Imran's test). I removed it: I didn't find bad consequences of that, but I'm not 100% sure why it was added in first place. I can imagine it was a desire to follow the proposed design for the Email project.

m1r0 reviewed (dismissed):
Thanks for working on this Dmitry. It works well in my opinion.

I've noticed that when you have a very wide image, the height is not good:

| Editor | Gmail/Preview |
| --- | --- |
| ![image](https://github.com/user-attachments/assets/a13d7baf-ac5f-4df0-8ba8-215e5d7fa7a5) | ![image](https://github.com/user-attachments/assets/08d17d27-1f6b-4038-8a78-e1690440ad82) |

You can see that in Gmail/Preview the pickle's height is bigger. Adding `(max-)height: 100%` fixes it.

The above is not a blocker and can be fixed in a separate PR.

merkushin committed:
Merge branch 'trunk' into fix/email-styles

merkushin committed:
Set 100% height for an image to avoid stretching of wide images

merkushin commented:
Thanks, @m1r0! I updated the PR, added 100% height for `<img/>` inside `.wp-image-block`.

m1r0 commented on assets/css/email-notifications/email-style.scss:
```diff
@@ -35,6 +35,10 @@ $MOBILE_HEADER_SIZE: 32px;
 }
 
+.wp-block-image img {
+	height: 100%;
```
I'm not sure if it makes a difference but probably `height: auto` is more correct. WDYT?

merkushin replied:
Thanks, Miro, I'll try it.

merkushin replied:
Yes, it works. Updated here: https://github.com/Automattic/sensei/pull/7595/commits/6a5e1a746994ec97c295c2033d169bc62b8a6caa

m1r0 reviewed (commented):
Sorry for the back and forth, but I've noticed that the image is missing a `box-sizing: border-box`. The issue is visible when you add a border to the image - it extends past the container width.

| Without `box-sizing: border-box` | With `box-sizing: border-box`  |
| --- | --- |
| ![image](https://github.com/user-attachments/assets/bf7fab0c-55cf-4e4d-80b0-c9643ae94710) | ![image](https://github.com/user-attachments/assets/11e83cc0-3d4a-4419-a0e5-14f8748ff4c4) |

merkushin committed:
Merge branch 'trunk' into fix/email-styles

merkushin committed:
Adjust height and box-rizing for image

merkushin commented:
Thanks @m1r0! I updated the PR. Could you look at it again? 😅 

![CleanShot 2024-07-28 at 13 28 35@2x](https://github.com/user-attachments/assets/255c557b-df9e-435e-a5cd-061ae8d9ea3d)

m1r0 reviewed (approved):
Looks good! Thanks for the updates. 🙏

merkushin committed:
Merge branch 'trunk' into fix/email-styles
